### ISSUE TITLE
[dx11][pre-ll] Partial updates of dynamic resources are broken

### ISSUE THREAD

**Issue by kvark** [gfx-rs/gfx#1418]
In D3D11, dynamic resources can only be mapped with either `NO_OVERWRITE` or `DISCARD` flags. The former only works on vertex/index buffer. The latter discards the previous contents, thus not supporting partial updates. This is what we do now, and it results in user data being dropped.

Problem-1: we do not detect/handle this properly. At the very least, we should return an error.
Problem-2: we need to re-evaluate the whole `Dynamic` usage (only for `pre-ll` branch). D3D11 doesn't actually allow partial updates to happen to GPU-visible resources (like GL does). You can only do this via staging resources, which can be mapped for writing, then copied to GPU-visible resources (basically, PBO path in GL).

cc @zakorgy @Bastacyclop

**Comment by kvark** [comment]
Thinking about this more, I think I got all the puzzle pieces. Core thing to realize was that the data buffer carried by `Encoder` is conceptually (and should be - factually) a PBO:
  - the encoder maps a PBO for the duration of command encoding
  - whenever the user issues `update_buffer`/`update_texture`:
    - we'll copy data data into the mapped region
    - we'll record the underlying `copy_buffer`/`copy_buffer_to_texture` call in the command buffer

This way, we don't rely on any discard mechanics in the backend, and we don't need a tricky `Dynamic` type usage at all. The only problem that comes to mind is that the `Encoder` doesn't know the size of the PBO it will need, and it doesn't have access to a factory for creating more of them or resizing. Thus, I believe the user will just need to specify the size of associated PBO when creating an encoder.

@msiglreith would you see any problems implementing this scheme in the `gfx_render` to work with LL?

**Comment by kvark** [comment]
That being said, we need an easier workaround for `pre-ll` branch. It's sad to have this broken on DX11.

**Comment by msiglreith** [comment]
> would you see any problems implementing this scheme in the gfx_render to work with LL?

It's actually also a `core` issue https://github.com/gfx-rs/gfx/issues/1416#issuecomment-320630776. Depending on if we want support in buffer updates.

For `render` this sounds reasonable in combination with a small buffer for the whole encoder, but the current `GraphicsEncoder` in `render` is only a temporary wrapper/encoder. Therefore we would need some changes here to handle a persitent scratch buffer.

**Comment by kvark** [comment]
More in-depth investigation here - https://github.com/gpuweb/gpuweb/issues/28

**Comment by Bastacyclop** [comment]
@kvark I never liked the `Dynamic` usage that was not part of my initial proposal and only reintroduced for `GL_DYNAMIC_STORAGE_BIT` and `D3D11_USAGE_DYNAMIC` (the broken part). I remember we already discussed the PBO solution at the time and the buffer size was not the only problem. Maybe it had something to do with the unability to keep the `Encoder` `Send` with it ?

I think I would rather have some kind of `UploadQueue` (a step towards higher level abstractions as discussed for [mapping overwrites](https://github.com/gfx-rs/gfx/issues/1334#issuecomment-310764144)) in `render` that is basically the PBO solution without bringing the latter inside the `Encoder`. It probably requires more fine-grained synchronisation but would reduce waste (multiple fences for a single buffer that can be used by multiple encoders), which shouldn't be a problem with LL (@msiglreith ?). For pre-LL it could just be a ring buffer of upload buffers with more waste.

**Comment by Bastacyclop** [comment]
Let's try to clarify the situation a bit more.

The proposed fix is to keep an `Upload` buffer inside the encoders, with a fixed size specified at creation.
I don't know what are the multithread constraints for D3D11 but with OpenGL we can't acquire (and release) the `Upload` mapping outside of the GL context thread.

**With pre-LL** this is a problem since we want encoders to be `Send`. I see two solutions:
- Ensure that the mapping is always available/acquired when the `Encoder ` is available (i.e. outside submission).
- Keep writing into an intermediary `Vec` before actually copying to an `Upload` buffer during submission.

The first solution requires less copies. Releasing the mapping before submission if necessary is not a problem, but we also need to ensure the mapping is available after submission which means eagerly synchronizing. This might not be a problem if this synchronisation is smart enough (e.g. we don't want to wait for a following draw command) and does not cost too much.

The second solution seems easier to implement and does not require an `Upload` buffer for each `Encoder`. This is the one I would pick for a simple pre-LL fixing.

We could completly remove the `Dynamic` usage or keep it if we don't want breaking changes, or to modify less backends. I also fear #1208 might complicate things for D3D11 texture updates.

**With LL** the encoders will be neither `Sync` nor `Send` so there should be no thread issue. However without `Send` there is little convenience provided by update commands IMO. So I would not even emulate them and instead provide better upload abstractions mentioned in the last comment.

@kvark What do you think about all this ?

**Comment by kvark** [comment]
@Bastacyclop Thank you for the analysis!

### pre-LL

We could also consider solution 3) map-write-unmap directly in the `Encoder::copy_*` implementation.

Overall, I wouldn't worry too much about pre-LL fixing. We haven't promised a "Long Term Support" kind of service for it anyway. Yes, it sucks that one of the basic things on D3D11 doesn't work properly. If there is an easy-ish workaround, let's do it! Otherwise - I prefer to focus on post-LL, but will gladly review the patches and publish new versions.

### post-LL

> So I would not even emulate them and instead provide better upload abstractions mentioned in the last comment.

Getting Vulkan Portability is one of the strategic goals of LL. If we don't provide `update_buffer` in core, we'll have to emulate it on all platforms by the portability shim around gfx-rs. This would be less efficient on Vulkan than to call native `update` methods, so is undesirable.

As to how exactly D3D12/Metal backends are going to do it:
  - `CommandPool` implementation will manage a grow-able pool of `Upload` buffers
  - when encoder needs to update something, it requests the buffer space from the active command pool
  - buffer space returns to the pool upon resetting the command buffer

**Comment by Bastacyclop** [comment]
> We could also consider solution 3) map-write-unmap directly in the Encoder::copy_* implementation.

This is not possible with OpenGL. I guess it is possible with D3D11 ?

**Comment by kvark** [comment]
@Bastacyclop yeah, our command buffers are backed by `ID3D11DeviceContext`, which can do `Map`/`Unmap`: https://msdn.microsoft.com/en-us/library/windows/desktop/ff476457(v=vs.85).aspx
